Add a post about maintenance (#1773)
rust-lang/blog.rust-lang.org: content/inside-rust/what-is-maintenance-anyway.md

This kind of "janitorial" work is quite challenging for several reasons:

- It can be difficult to demonstrate the positive effects of this work on a given software project. Users sometimes assume that if something works now, it will keep working indefinitely. Furthermore, Rust has a very high bar for being reliable (in many aspects of that word), so people are used to the fact that the compiler works correctly, that critical bugs are resolved as quickly as possible, that new releases are published every six weeks, etc. But all that takes considerable effort, even without considering shipping any new features. Therefore, a successful report of maintenance work can be simply "things still work as they used to".
- It can be very unpredictable, because it is hard to estimate in advance what things will break next. This also makes it difficult to plan this work ahead and make promises about what kind of work will get done in a given time period.
- Finding motivation for performing a lot of maintenance work can be a struggle. It is often not work that someone *wants* to do, but work that *has to be done*. That can be draining, especially if that work is done by volunteers.

> 

Reviewer: ```suggestion
- Maintenance is often not on a roadmap, and often not viewed as high-status work that gets someone recognized or promoted. (This relates to it being characterized by an absence of noteworthy incidents.)

```

## Enabling evolution

> The maintenance activities described above are required to keep a software project working. However, usually that's not *all* we want. Most software projects have a tendency to *evolve* over time, which usually means adding new features and making improvements. Improving the language, compiler, standard library, tooling, etc. might unblock some Rust users, which is great. But what might not be so obvious is that we often need to unblock the contributors who implement these new features in the first place!

Reviewer: ```suggestion
The maintenance activities described above are required to keep a software project working. However, usually that's not *all* we need. Most software projects need to *evolve* over time, and Rust is no exception; this usually means adding new features and making improvements. Improving the language, compiler, standard library, tooling, etc. might unblock some Rust users, which is great. But what might not be so obvious is that we often need to unblock the contributors who implement these new features in the first place!
```

Before someone even starts to implement a new feature, they often want to get a vibe check from someone else knowledgeable with the corresponding part of the codebase, to consult how best to approach the implementation. Furthermore, certain features require complex refactorings to be performed before they can be implemented. Then, once a pull request with a new feature is submitted, someone has to review it. And even once a pull request lands, it can introduce new tech debt, which eventually has to be paid by future refactorings.

> However, it's not only refactoring and code reviews that help make progress on various improvements. Even implementing a new feature can unblock many other contributors, who might want to add other features that depend on it.

Reviewer: ```suggestion
However, it's not only refactoring and code reviews that help make progress on various improvements. Even implementing a new feature can unblock many other contributors, who might want to add other features that depend on it. Implementing a new feature may also unblock new use cases for many Rust users, who can then give us valuable feedback on other fixes or improvements for the future.
```

Trying to capture one aspect of how the addition of new features is important for Rust to remain relevant and useful and have users.

## Enabling evolution

The maintenance activities described above are required to keep a software project working. However, usually that's not *all* we need. Most software projects need to *evolve* over time, and Rust is no exception; this usually means adding new features and making improvements. Improving the language, compiler, standard library, tooling, etc. might unblock some Rust users, which is great. But what might not be so obvious is that we often need to unblock the contributors who implement these new features in the first place!

Before someone even starts to implement a new feature, they often want to get a vibe check from someone else knowledgeable with the corresponding part of the codebase, or advice on how best to approach the implementation. Furthermore, certain features require complex refactorings to be performed before they can be implemented. Then, once a pull request with a new feature is submitted, someone has to review it. And even once a pull request lands, it may introduce new [technical debt](https://en.wikipedia.org/wiki/Technical_debt), which will eventually need cleaning up.

> However, it's not only refactoring and code reviews that help make progress on various improvements. Even implementing a new feature can unblock many other contributors, who might want to add other features that depend on it. Implementing a new feature may also unblock new use cases for many Rust users, who can then give us valuable feedback on other fixes or improvements for the future.

Reviewer 2: I think the last sentence is true, but I also don't think it's relevant to this section. This falls more squarely into the realm of "shipping features" which is outside the realm of pure maintenance work.

(I think individual maintainers should be doing some of both, but pure feature work still falls outside my definition of maintenance work.)

Reviewer: > This falls more squarely into the realm of "shipping features" which is outside the realm of pure maintenance work.

I think this is something where reasonable people can disagree. I personally think the ongoing work of meeting user needs can be part of maintenance, and would use a more expansive definition of "maintenance" and "maintainer".

Broadly speaking, I feel the term "maintainer" has evolved to a more general "the people who own an Open Source project" (as distinguished from "contributor"). This isn't a perfect definition, but to a zeroth approximation I would happily define "maintainer" as "everyone with r+ or similar recognized privileges and any form of (partial, shared) collective ownership".

I also think this sentence, in particular, feeds nicely into the next paragraph.

Reviewer 2: I can see such a broad definition being useful in some contexts, just not when it comes to funding tied to specific outcomes. I think funders ought to know what they are "buying". If the answer is "it can include anything the maintainer wants to work on, as long as it can be said to meet the needs of some user", that's a much different answer than "work that keeps the lights on and enables new development work to proceed". And I do think it's important to keep this distinction clear, even in cases where the same person and the same funding support both maintenance and feature-driven work, so we can communicate more effectively.

While I do think reasonable people can disagree on what definitions are useful, my reading of the post is that it is meant to reflect what our group all agrees is included in the definition of maintenance. So I would suggest removing the sentence from this post while the point is under discussion.

Author: Hmm, it was certainly not my intention to anyhow "cement" what is the only definition of maintenance. The post was originally targeted to e.g. Lori, who kept asking "what is maintenance", so I wanted to offer my point of view and explain how I see it, with the hope that it will be universal enough and non-controversial.

I think that the post could be useful as a frame of reference if someone asks what even is maintanenance, so that we can refer them to this post to get an idea. But I certainly don't want anyone to go "wait, why are you doing X, that is clearly not maintenance, it was not mentioned in this post", or something. It's just a blog post...

Reviewer: If the purpose is to explain what people mean when they say "maintenance" or "maintainer", then it seems useful to explain that different people have different connotations for the terms.

And all of that is another part of what maintainers do. They continuously improve the codebase so that it is easier to land new changes, they review pull requests, and they communicate with and mentor other contributors. This work is crucial to ensure the long-term health of constantly evolving codebases. And it also has a strong second-order effect. By unblocking other Rust contributors, new features and improvements can be landed quicker and more frequently, which in turns unblocks and improves the lives of Rust users. That is a multiplicative effect that strongly benefits everyone!

This enablement work of course requires having a deep knowledge of the codebase, but also a long-term focus to perform complex refactorings and grit to continuously review tons of pull requests.

## Conclusion

Based on the observations presented above, we could define a maintainer as someone who continuously ensures that a software project keeps working, but who also does a lot of hard (and often invisible) work to enable other contributors to evolve and improve the project.

> You probably already guessed that maintenance work is not easy, and maybe that is one of the reasons why there seems to be a perpetual imbalance. There are often many people who want to contribute to open-source, by making improvements and implementing new features, as it is often seen as being fun, and implementing a new feature is also a very visible achievement that people can brag about. But there are way fewer people who want to continuously maintain a codebase (especially if they are not paid for it!), as the results of good maintenance are much more difficult to demonstrate, and it more often than not resembles work than pure fun.

Reviewer: ```suggestion
You probably already guessed that maintenance work is not easy, and maybe that is one of the reasons why there seems to be a perpetual imbalance. There are often many people who want to contribute to open source, by making improvements and implementing new features, as it is often seen as being fun, and implementing a new feature is also a very visible achievement that may help meet someone's needs or give them something obvious to be proud of. But there are far fewer people who want to continuously maintain a codebase (especially if they are not paid for it!), as the results of good maintenance are much more difficult to demonstrate, and it often more closely resembles work rather than pure fun.
```

One tweak towards wording with less negative connotations (brag -> be proud of), and some copyedits.

While being a maintainer can be challenging, *becoming* one is even more difficult. It is not really possible to become a maintainer overnight, as it takes time to gain deep expertise in a certain part of a codebase, but also to gain trust of other maintainers and contributors. All while continuously doing work that no one else wants to do, as it is often not very exciting.

> We are incredibly happy that we have so many awesome maintainers in the Rust Project, who pour their souls into ensuring that Rust becomes better every day, and who invested enormous amounts of time into becoming experts in certain Rust projects so that they could effectively maintain them. However, many of those maintainers are volunteers, and when so much maintenance burden falls on someone who still has to work another job to pay the bills, it can lead to burnout. And that is something that we want to prevent.

Reviewer: ```suggestion
We are incredibly happy that we have so many awesome maintainers in the Rust Project, who pour their souls into ensuring that Rust becomes better every day, and who invest enormous amounts of time into becoming experts in certain Rust projects so that they can effectively maintain them. However, many of those maintainers are volunteers, and when so much maintenance burden falls on someone who still has to work another job to pay the bills, it can lead to burnout. This is something that we want to prevent.
```

Matching tense, and other minor copyedits.
